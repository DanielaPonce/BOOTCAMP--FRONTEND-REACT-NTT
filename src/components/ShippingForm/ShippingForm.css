/* ESTILOS PARA SHIPPING FORM */
.shipping-form {
	width: 100%;
	border-radius: 4px;
	padding: 24px;
	background-color: var(--white);
	display: flex;
	flex-direction: column;
	gap: 1rem;
}

.shipping-form h2 {
	margin: 0 auto;
}

form .form-group {
	display: flex;
	flex-direction: column;
	gap: 4px;
}

.form-group input, select {
	border: 1px solid var(--border-color);
	padding: 0.7rem;
	font-size: 1rem;
	border-radius: 4px;
}

.form-group input:focus {
	outline: none;
	border-color: var(--text-color);
}

.form-group .error-message {
	transition: all 0.3s;
	color: var(--error-color);
}

form button {
	background-color: var(--secondary-color);
	color: var(--white);
	padding: 0.7rem 1.5rem;
	border: none;
	border-radius: 5px;

	font-size: 16px;
	font-weight: 600;
	cursor: pointer;
	transition: background-color 0.3s;

	margin: 0 auto;
	min-width: 24rem;
}

@media only screen and (max-width: 739px) { 
	form button {
		min-width: 100%;
	}
}