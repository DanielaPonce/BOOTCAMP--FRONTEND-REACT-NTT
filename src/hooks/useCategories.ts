import { useEffect, useState } from 'react';
import { fetchAllCategories } from '../api/category.api';
import { Category } from '../models/category.types';

export const useCategories = () => {
	const [categories, setCategories] = useState<Category[]>([]);

	useEffect(() => {
		fetchAllCategories().then((data) => setCategories(data));
	}, []);

	return { categories };
};
