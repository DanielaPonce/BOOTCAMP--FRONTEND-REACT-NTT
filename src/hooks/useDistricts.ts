import { useEffect, useState } from 'react';

interface District {
	id: number;
	name: string;
}

export const useDistricts = () => {
	const [districts, setDistricts] = useState<District[]>([]);

	useEffect(() => {
		const fetchDistricts = async () => {
			const response = await fetch('/src/assets/mock/districts.json');
			const data: District[] = await response.json();
			setDistricts(data);
		};

		fetchDistricts();
	}, []);

	return districts;
};
