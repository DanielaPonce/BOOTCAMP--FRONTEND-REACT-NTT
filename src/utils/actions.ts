export enum Actions {
	ADD_TO_CART_TYPE = 'ADD_TO_CART',
	REMOVE_FROM_CART_TYPE = 'REMOVE_FROM_CART',
	UPDATE_QUANTITY_TYPE = 'UPDATE_QUANTITY',
	CLEAR_CART_TYPE = 'CLEAR_CART'
}
