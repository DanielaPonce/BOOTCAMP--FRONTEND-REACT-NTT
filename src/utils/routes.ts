export enum RoutesConstants {
	home = '/',
	summary = 'summary'
}
