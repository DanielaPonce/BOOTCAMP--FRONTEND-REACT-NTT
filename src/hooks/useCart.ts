import { useEffect } from 'react';
import { useCartContext } from '../context/CartContext';
import { Product } from '../models/product.types';
import { Actions } from '../utils/actions';
import { LocalStorageConstants } from '../utils/localStorageConstants';

export const useCart = () => {
	const { state, dispatch } = useCartContext();

	useEffect(() => {
		localStorage.setItem(
			LocalStorageConstants.CartKey,
			JSON.stringify(state)
		);
	}, [state]);

	const addToCart = (product: Product) =>
		dispatch({ type: Actions.ADD_TO_CART_TYPE, payload: product });

	const removeFromCart = (id: number) =>
		dispatch({ type: Actions.REMOVE_FROM_CART_TYPE, payload: id });

	const updateQuantity = (id: number, quantity: number) =>
		dispatch({
			type: Actions.UPDATE_QUANTITY_TYPE,
			payload: { id, quantity }
		});

	const clearCart = () => dispatch({ type: Actions.CLEAR_CART_TYPE });

	return {
		cart: state.cart,
		total: state.total,
		itemCount: state.itemCount,
		addToCart,
		removeFromCart,
		updateQuantity,
		clearCart
	};
};
