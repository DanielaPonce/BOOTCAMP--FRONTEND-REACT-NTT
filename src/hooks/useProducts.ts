import { useEffect, useState } from 'react';
import { fetchAllProducts, fetchProductsByCategory } from '../api/product.api';
import { Product } from '../models/product.types';
import { Constants } from '../utils/constants';

export const useProducts = (category: string, search: string) => {
	const [allProducts, setAllProducts] = useState<Product[]>([]);
	const [filteredProducts, setFilteredProducts] = useState<Product[]>([]);
	const [isLoading, setIsLoading] = useState(true);
	const [currentPage, setCurrentPage] = useState(1);

	const itemsPerPage = 8;

	const loadProducts = async () => {
		setIsLoading(true);
		try {
			let products: Product[] = [];
			if (category === Constants.defaultSelectValue) {
				products = await fetchAllProducts();
			} else {
				products = await fetchProductsByCategory(category);
			}
			setAllProducts(products);
			setFilteredProducts(products);
			setCurrentPage(1);
		} catch (error) {
			console.log('Error al cargar los productos:', error);
		} finally {
			setIsLoading(false);
		}
	};

	useEffect(() => {
		loadProducts();
	}, [category]);

	useEffect(() => {
		const filtered = allProducts.filter((product) =>
			product.title.toLowerCase().includes(search.toLowerCase())
		);
		setFilteredProducts(filtered);
		setCurrentPage(1);
	}, [search, allProducts]);

	const paginatedProducts = filteredProducts.slice(
		(currentPage - 1) * itemsPerPage,
		currentPage * itemsPerPage
	);

	const totalPages = Math.ceil(filteredProducts.length / itemsPerPage);

	return {
		products: paginatedProducts,
		isLoading,
		totalPages,
		currentPage,
		setCurrentPage,
		showPagination: totalPages !== 1
	};
};
