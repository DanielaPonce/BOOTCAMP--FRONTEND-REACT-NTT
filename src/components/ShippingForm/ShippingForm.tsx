import { FC, useState } from 'react';
import { useDistricts } from '../../hooks/useDistricts';
import { useCart } from '../../hooks/useCart';
import { useNavigate } from 'react-router';
import { RoutesConstants } from '../../utils/routes';
import Button from '../Button/Button';
import { RegexPatterns } from '../../utils/regexPatterns';
import './ShippingForm.css';
interface FormState {
	firstName: string;
	lastName: string;
	district: string;
	address: string;
	reference: string;
	phoneNumber: string;
}

type ErrorsType = Partial<Record<keyof FormState, string>>;

const ShippingForm: FC = () => {
	const navigate = useNavigate();

	const { clearCart } = useCart();
	const districts = useDistricts();

	const [formState, setFormState] = useState<FormState>({
		firstName: '',
		lastName: '',
		district: '',
		address: '',
		reference: '',
		phoneNumber: ''
	});

	const [errors, setErrors] = useState<ErrorsType>({});

	const handleInputChange = (
		e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
	) => {
		const { name, value } = e.target;
		setFormState({ ...formState, [name]: value });

		if (errors[name as keyof FormState]) {
			setErrors({ ...errors, [name]: undefined });
		}
	};

	const validate = (): boolean => {
		const newErrors: ErrorsType = {};

		if (
			!formState.firstName.trim() ||
			!RegexPatterns.OnlyLetters.test(formState.firstName)
		) {
			newErrors.firstName = 'Debe ingresar un valor válido';
		}

		if (
			!formState.lastName.trim() ||
			!RegexPatterns.OnlyLetters.test(formState.lastName)
		) {
			newErrors.lastName = 'Debe ingresar un valor válido';
		}

		if (!formState.district.trim()) {
			newErrors.district = 'Campo obligatorio';
		}

		if (!formState.address.trim()) {
			newErrors.address = 'Campo obligatorio';
		}

		if (!formState.reference.trim()) {
			newErrors.reference = 'Campo obligatorio';
		}

		if (!formState.phoneNumber.trim()) {
			newErrors.phoneNumber = 'Campo obligatorio';
		} else if (!RegexPatterns.PhoneNumber.test(formState.phoneNumber)) {
			newErrors.phoneNumber = 'Debe ser un número de 9 dígitos';
		}

		setErrors(newErrors);
		return Object.keys(newErrors).length === 0;
	};

	const handleSubmit = (e: React.FormEvent) => {
		e.preventDefault();
		if (validate()) {
			alert('¡Su pedido se registró con éxito!');
			clearCart();
			navigate(RoutesConstants.Home);
		}
	};

	return (
		<form onSubmit={handleSubmit} className="shipping-form">
			<h2>Información de Envío</h2>

			<div className="form-group">
				<label htmlFor="firstName">Nombres</label>
				<input
					type="text"
					id="firstName"
					name="firstName"
					value={formState.firstName}
					onChange={handleInputChange}
					placeholder="Ingresa tus nombres"
				/>
				{errors.firstName && (
					<span className="error-message">{errors.firstName}</span>
				)}
			</div>

			<div className="form-group">
				<label htmlFor="lastName">Apellidos</label>
				<input
					type="text"
					id="lastName"
					name="lastName"
					value={formState.lastName}
					onChange={handleInputChange}
					placeholder="Ingresa tus apellidos"
				/>
				{errors.lastName && (
					<span className="error-message">{errors.lastName}</span>
				)}
			</div>

			<div className="form-group">
				<label htmlFor="district">Distrito</label>
				<select
					id="district"
					name="district"
					value={formState.district}
					onChange={handleInputChange}
				>
					<option value="">Selecciona tu distrito</option>
					{districts.map((district) => (
						<option key={district.id} value={district.name}>
							{district.name}
						</option>
					))}
				</select>
				{errors.district && (
					<span className="error-message">{errors.district}</span>
				)}
			</div>

			<div className="form-group">
				<label htmlFor="address">Dirección</label>
				<input
					type="text"
					id="address"
					name="address"
					value={formState.address}
					onChange={handleInputChange}
					placeholder="Ingresa tu dirección"
				/>
				{errors.address && (
					<span className="error-message">{errors.address}</span>
				)}
			</div>

			<div className="form-group">
				<label htmlFor="reference">Referencia</label>
				<input
					type="text"
					id="reference"
					name="reference"
					value={formState.reference}
					onChange={handleInputChange}
					placeholder="Referencia de tu dirección"
				/>
				{errors.reference && (
					<span className="error-message">{errors.reference}</span>
				)}
			</div>

			<div className="form-group">
				<label htmlFor="phoneNumber">Celular</label>
				<input
					type="text"
					id="phoneNumber"
					name="phoneNumber"
					value={formState.phoneNumber}
					onChange={handleInputChange}
					placeholder="Ingresa tu número de celular"
				/>
				{errors.phoneNumber && (
					<span className="error-message">{errors.phoneNumber}</span>
				)}
			</div>

			<Button type="submit" className="submit-button">
				Comprar
			</Button>
		</form>
	);
};

export default ShippingForm;
