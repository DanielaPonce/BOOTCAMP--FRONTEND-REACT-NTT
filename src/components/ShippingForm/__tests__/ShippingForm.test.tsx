import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { jest } from '@jest/globals';
import { useNavigate } from 'react-router';
import ShippingForm from '../ShippingForm';
import { RoutesConstants } from '../../../utils/routes';
import { useCart } from '../../../hooks/useCart';
import { useDistricts } from '../../../hooks/useDistricts';

jest.mock('../../../hooks/useCart.ts');
jest.mock('../../../hooks/useDistricts');
jest.mock('react-router', () => ({
	useNavigate: jest.fn()
}));

describe('ShippingForm Component', () => {
	const mockClearCart = jest.fn();
	const mockNavigate = jest.fn();
	const mockDistricts = [
		{ id: 1, name: 'San Isidro' },
		{ id: 2, name: 'Miraflores' }
	];

	beforeEach(() => {
		(useCart as jest.Mock).mockReturnValue({ clearCart: mockClearCart });
		(useDistricts as jest.Mock).mockReturnValue(mockDistricts);
		(useNavigate as jest.Mock).mockReturnValue(mockNavigate);
		mockClearCart.mockClear();
		mockNavigate.mockClear();
	});

	beforeAll(() => {
		jest.spyOn(window, 'alert').mockImplementation(() => {});
	});

	test('renders all form elements', () => {
		render(<ShippingForm />);

		expect(screen.getByLabelText('Nombres')).toBeInTheDocument();
		expect(screen.getByLabelText('Apellidos')).toBeInTheDocument();
		expect(screen.getByLabelText('Distrito')).toBeInTheDocument();
		expect(screen.getByLabelText('Dirección')).toBeInTheDocument();
		expect(screen.getByLabelText('Referencia')).toBeInTheDocument();
		expect(screen.getByLabelText('Celular')).toBeInTheDocument();

		expect(
			screen.getByRole('button', { name: /comprar/i })
		).toBeInTheDocument();
	});

	test('displays validation errors for required fields', () => {
		render(<ShippingForm />);

		const button = screen.getByRole('button', { name: /comprar/i });

		fireEvent.click(button);

		expect(
			screen.getAllByText('Debe ingresar un valor válido').length
		).toBeGreaterThan(0);
		expect(screen.getAllByText('Campo obligatorio').length).toBe(4);
	});

	test('updates form state on input change', () => {
		render(<ShippingForm />);

		const firstNameInput = screen.getByLabelText('Nombres');
		fireEvent.change(firstNameInput, { target: { value: 'Carlos' } });

		expect(firstNameInput).toHaveValue('Carlos');
	});

	test('validates incorrect phone number format', () => {
		render(<ShippingForm />);

		const phoneInput = screen.getByLabelText('Celular');
		fireEvent.change(phoneInput, { target: { value: '12345' } });

		const button = screen.getByRole('button', { name: /comprar/i });
		fireEvent.click(button);

		expect(
			screen.getByText('Debe ser un número de 9 dígitos')
		).toBeInTheDocument();
	});

	test('submits form with valid data', () => {
		render(<ShippingForm />);

		fireEvent.change(screen.getByLabelText('Nombres'), {
			target: { value: 'Carlos' }
		});
		fireEvent.change(screen.getByLabelText('Apellidos'), {
			target: { value: 'Pérez' }
		});
		fireEvent.change(screen.getByLabelText('Distrito'), {
			target: { value: 'San Isidro' }
		});
		fireEvent.change(screen.getByLabelText('Dirección'), {
			target: { value: 'Av. Principal 123' }
		});
		fireEvent.change(screen.getByLabelText('Referencia'), {
			target: { value: 'Cerca al parque' }
		});
		fireEvent.change(screen.getByLabelText('Celular'), {
			target: { value: '987654321' }
		});

		const button = screen.getByRole('button', { name: /comprar/i });
		fireEvent.click(button);

		expect(
			screen.queryAllByText('Debe ingresar un valor válido').length
		).toBe(0);
		expect(screen.queryAllByText('Campo obligatorio').length).toBe(0);

		expect(window.alert).toHaveBeenCalledWith(
			'¡Su pedido se registró con éxito!'
		);

		expect(mockClearCart).toHaveBeenCalledTimes(1);
		expect(mockNavigate).toHaveBeenCalledWith(RoutesConstants.Home);
	});

	test('does not submit form with invalid data', () => {
		render(<ShippingForm />);

		fireEvent.change(screen.getByLabelText('Nombres'), {
			target: { value: '' }
		});
		const button = screen.getByRole('button', { name: /comprar/i });
		fireEvent.click(button);

		expect(mockClearCart).not.toHaveBeenCalled();
		expect(mockNavigate).not.toHaveBeenCalled();
	});

	test('matches snapshot', () => {
		const { asFragment } = render(<ShippingForm />);
		expect(asFragment()).toMatchSnapshot();
	});
});
