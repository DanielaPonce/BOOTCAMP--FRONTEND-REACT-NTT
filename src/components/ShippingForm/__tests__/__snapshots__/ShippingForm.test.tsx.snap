// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShippingForm Component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="shipping-form"
  >
    <h2>
      Información de Envío
    </h2>
    <div
      class="form-group"
    >
      <label
        for="firstName"
      >
        Nombres
      </label>
      <input
        id="firstName"
        name="firstName"
        placeholder="Ingresa tus nombres"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="lastName"
      >
        Apellidos
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Ingresa tus apellidos"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="district"
      >
        Distrito
      </label>
      <select
        id="district"
        name="district"
      >
        <option
          value=""
        >
          Selecciona tu distrito
        </option>
        <option
          value="San Isidro"
        >
          San Isidro
        </option>
        <option
          value="Miraflores"
        >
          Miraflores
        </option>
      </select>
    </div>
    <div
      class="form-group"
    >
      <label
        for="address"
      >
        Dirección
      </label>
      <input
        id="address"
        name="address"
        placeholder="Ingresa tu dirección"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="reference"
      >
        Referencia
      </label>
      <input
        id="reference"
        name="reference"
        placeholder="Referencia de tu dirección"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="phoneNumber"
      >
        Celular
      </label>
      <input
        id="phoneNumber"
        name="phoneNumber"
        placeholder="Ingresa tu número de celular"
        type="text"
        value=""
      />
    </div>
    <button
      class="btn submit-button "
      type="submit"
    >
      Comprar
    </button>
  </form>
</DocumentFragment>
`;
